File: 1244-DesignALeaderboard/1244-DesignALeaderboard.java
// Last updated: 1/12/2026, 10:13:59 PM
1class Leaderboard {
2    private Map<Integer, Integer> playerScores;
3    private TreeMap<Integer, Integer> sortedScores;
4
5    public Leaderboard() {
6        playerScores = new HashMap<>();
7        sortedScores = new TreeMap<>();
8    }
9
10    public void addScore(int playerId, int score) {
11        if (playerScores.containsKey(playerId)) {
12            removeScore(playerScores.get(playerId));
13        }
14        int newScore = playerScores.getOrDefault(playerId, 0) + score;
15        playerScores.put(playerId, newScore);
16
17        sortedScores.put(newScore, sortedScores.getOrDefault(newScore, 0) + 1);
18    }
19
20    public int top(int K) {
21        int sum = 0;
22        int count = 0;
23
24        for (int score : sortedScores.descendingKeySet()) {
25            int times = sortedScores.get(score);
26
27            int take = Math.min(times, K - count);
28            sum += take * score;
29            count += take;
30
31            if (count == K)
32                break;
33        }
34        return sum;
35    }
36
37    public void reset(int playerId) {
38        int score = playerScores.get(playerId);
39        removeScore(score);
40        playerScores.remove(playerId);
41    }
42
43    private void removeScore(int score) {
44        sortedScores.put(score, sortedScores.get(score) - 1);
45        if (sortedScores.get(score) == 0) {
46            sortedScores.remove(score);
47        }
48    }
49}
50
51/**
52 * Your Leaderboard object will be instantiated and called as such:
53 * Leaderboard obj = new Leaderboard();
54 * obj.addScore(playerId,score);
55 * int param_2 = obj.top(K);
56 * obj.reset(playerId);
57 */

File: 1244-DesignALeaderboard/1244-DesignALeaderboard.py
# Last updated: 1/12/2026, 9:38:32 PM
1from sortedcontainers import SortedList
2
3class Leaderboard:
4
5    def __init__(self):
6        # 字典：玩家ID → 当前总分
7        self.player_scores = {}
8        
9        # 有序列表：存储所有分数（自动升序排列）
10        self.sorted_scores = SortedList()
11
12    def addScore(self, playerId: int, score: int) -> None:
13        if playerId in self.player_scores: 
14            # 1. 从有序列表中删除旧分数
15            self.sorted_scores.remove(self.player_scores[playerId])
16            
17            # 2. 更新玩家总分
18            self.player_scores[playerId] += score 
19        else: 
20            # 新玩家：直接记录分数
21            self.player_scores[playerId] = score
22        
23        # 3. 将新分数插入有序列表（自动保持排序）
24        self.sorted_scores.add(self.player_scores[playerId])
25
26    def top(self, K: int) -> int:
27        # 取最后 K 个分数（最大的 K 个）并求和
28        return sum(self.sorted_scores[-K:])
29
30    def reset(self, playerId: int) -> None:
31        # 1. 从有序列表中删除该玩家的分数
32        self.sorted_scores.remove(self.player_scores[playerId])
33        
34        # 2. 从字典中删除该玩家
35        self.player_scores.pop(playerId)